---
title: "Chip-seq in Galaxy and R"
author: "Emily Chambers, Mark Dunning"
date: '`r format(Sys.time(), "Last modified: %d %b %Y")`'
output: 
  html_notebook: 
    toc: yes
    toc_float: yes
editor_options: 
  chunk_output_type: inline
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,message=FALSE)
```
# Hands-on chip-seq Analysis in Galaxy and R

### Sheffield Bioinformatics Core
<img src="images/logo-sm.png" align=right>

web : [sbc.shef.ac.uk](http://sbc.shef.ac.uk)  
twitter: [SheffBioinfCore](https://twitter.com/SheffBioinfCore)  
email: [bioinformatics-core@sheffield.ac.uk](bioinformatics-core@sheffield.ac.uk)

-----

# Acknowledgement

Some of these materials are based on courses available at .

- [ATAC-Seq data analysis in Galaxy](https://training.galaxyproject.org/training-material/topics/epigenetics/tutorials/atac-seq/tutorial.html/)
- [Chip-seq excersizes in R by Jonas Ibn-Salem, JGU Mainz](http://cbdm-01.zdv.uni-mainz.de/~jibnsale/teaching/chip-seq_exercises.html)

-----

# Tutorial overview
This tutorial will cover the basics of processing a chip-seq analysis using Galaxy; a open-source web-based platform for the analysis of biological data. You should gain an appreciation of the tasks involved in a typical NGS analysis and be comfortable with the outputs generated by a sequencing service.

This morning tutorial covers the following steps in our analysis pipeline


## Introduction to chip-seq

What is chip-seq data?

![](images/szalkowski_schmid_2011.jpg)

## The data

The data for this tutorial are publicly-available 

-----

# Section 1: Preparation


**Ignore if you have already created a Galaxy account and uploaded the example data in a previous exercise**

#### 1.  Register as a new user on one of the public Galaxy servers

- https://usegalaxy.org
- https://usegalaxy.org.au
- https://usegalaxy.eu

**Make sure you check your email to activate your account**

#### 2.  Import the data for the workshop.

We can going to import the [*fastq* files](https://en.wikipedia.org/wiki/FASTQ_format) for this experiment. This is a standard format for storing raw sequencing reads and their associated quality scores. However, as we will see, the representation of the quality scores has changed over time.

You can import the data by:

1.  In the tool panel located on the left, under Basic Tools select **Get
    Data > Upload File**. Click on the **Paste/Fetch data** button on the
    bottom section of the pop-up window.
2.  Upload the sequence data by selecting the files `JoeBlogsBRCAPanel_R1.fastq` and `JoeBlogsBRCAPanel_R2.fastq`. You don't need to specify the file type or genome build. Galaxy should be able to make a reasonable guess.


3.  You should now have these 2 files in your history:
    - `JoeBlogsBRCAPanel_R1.fastq`
    - `JoeBlogsBRCAPanel_R2.fastq`
    
    # Section 2: Fastq file format

You can view the files you just uploaded by clicking the **eye icon** the history item. The first few lines should read as follows


**JoeBlogsBRCAPanel_R1.fastq**

```
@HWI-D00461:188:HVGY2BCXY:1:1101:1363:84148/1
TGTGTCATTTCTATTATCTTTGGAACAACCATGAATTAGTCCCTTGGGGTTTTCAAATGCTGCACACTGACTCACACATTTATTTGGTTCTGTTTTTGCCTTCCCTNN
+
DDDDDIHIIIIIHIIIIIIIIIIHIIIIIIIIIHIIIIIIHHIIIIIIIHIIIHIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIHHIDHHIHC#
```

**JoeBlogsBRCAPanel_R2.fastq**

```
@HWI-D00461:188:HVGY2BCXY:1:1101:1363:84148/2
TGGAAAGACTTTTGGGGGGGGGAGTATTTTTCTTGTTTCTGGTTTTGGTTTTTTTGATCCGGGAAAGATTTTGTTTTTTGGAGGTTGGACTTTTGGGGAGGGGAAAAN
+
<00<<1111<1<11/0/<///</<0111DF11<<11111<11111<1/1<1D1///<<11<<</<1111<<11<DD1<//<110<11/11<0<0</0<0<<-//<<FE
```


The first line is the unique identifier for each sequenced read. It can be used to encode information such as the *sequencing machine*, *flow cell* and *lane* that the read was generated from and the physical coordinates on the lane.  

The quality scores are [ASCII](http://ascii-code.com/) representations of how confident we are that a particular base has been called correctly. Letters that are further along the alphabet indicate higher confidence. This is important when trying to identify types of genome variation such as single base changes, but is also indicative of the overall quality of the sequencing. Different scales have been employed over time (resulting in a different set of characters appearing in the file). 


### Deriving the Quality Score

First of all, we convert the base-calling probability (p) into a `Q` score using the formula

- Quality scores $$ Q = -10log_{10}p$$
    + Q = 30, p=0.001
    + Q = 20, p=0.01
    + Q = 10, p=0.1
- These numeric quantities are *encoded* as [**ASCII**](http://ascii-code.com/) code
    + At least 33 to get to meaningful characters
    (https://en.wikipedia.org/wiki/FASTQ_format)
    
![](images/phred.png)      

### Quality Scores to probabilities

- look-up the ASCII code for each character
- subtract the offset to get the Q score
- convert to a probability using the formula:-

$$ p = 10^{-Q/10} $$

Let's see this calculation for the first few bases of the first read in `JoeBlogsBRCAPanel_R1.fastq`; `DDDDDIHI....`

Character  | Code | Minus 33 Offset | Probability
------------- | ------------- | ------------- | -------------
D  | 68 | 35 | 0.0003162278
D  | 68 | 35 | 0.0003162278
D  | 68 | 35 | 0.0003162278
D  | 68 | 35 | 0.0003162278
D  | 68 | 35 | 0.0003162278
I  | 73 | 40 | 0.0001000000
H  | 72 | 39 | 0.0001258925
I  | 73 | 40 | 0.0001000000

In practice, we don't have to convert the values as we have software that will do this automatically

-----

# Section 3: Quality assessment with FastQC

[FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/) is a popular tool from [Babraham Institute Bioinformatics Group](https://www.bioinformatics.babraham.ac.uk/index.html) used for *quality assessment* of sequencing data. Most Bioinformatics pipelines will use FastQC, or similar tools in the first stage of the analysis. The [documentation](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/) for FastQC will help you to interpret the plots and stats produced by the tool. A traffic light system is used to alert the user's attention to possible issues. 

- From the left hand tool panel in Galaxy, under *GENOMIC FILE MANIPULATION*, select *FASTQ Quality Control -> FastQC*
- Select one of the FASTQ files as input and *Execute* the tool.
- When the tool finishes running, you should have an HTML file in your History. Click on the eye icon to view the various quality metrics.

The most important image is whether the base quality decreases significantly over the length of the read

![](images/fastqc-2a.png)

Good quality data should look something like:-

![](images/fastqc-2b.png)



Look at the generated FastQC metrics for your uploaded fastq files. This data looks pretty good - high per-base quality scores (most above 30).

All is not lost if we observe poor quality bases towards the end of the read. There are a number of *trimming* options that we can use for NGS data and some of these are available through Galaxy. Check out the [Trimmming Reads](https://galaxyproject.org/tutorials/ngs/#trimming-reads) section of the Galaxy NGS tutorial if you are interested in how we can trim our reads.

It is also worth bearing in mind that the tool is blind to the particular type of sequencing you are performing (i.e. whole-genome, ChIP-seq, RNA-seq) and the organism being sequenced, so some warnings might be expected due to the nature of your experiment. For instance, there are known sequencing composition biases that can occur at the beginning of RNA-seq reads. 

## Aggregating QC reports with multiqc

For datasets with large numbers of fastq files, it may be useful to aggregate the individual reports into a single combined report. 

- Under *GENOMIC FILE MANIPULATION*, select *FASTQ Quality Control -> MultiQC*
- Make sure *Software name* is set to `FastQC`
- In *Results file*, select the **RawData** results files that you have just generated

You should then be able to view the fastqc plots for both the fastq files on the same page.

# Section 4: Alignment

We don't really spend much time look at *fastq* files, as most of our time is spent with *aligned* reads. i.e. we have used some software to tell us whereabouts in the genome each read belongs to. This will *usually* be performed for you as part of a sequencing service, but it is good to get an appreciation of the steps involved.

In this section we map the reads in our FASTQ files to a reference genome. 

A plethora of different tools have been written to perform this task, and we will not describe it in detail. Links to some key publications are given below:-

  + 2009 Bowtie 1 - [Langmead et al](http://genomebiology.com/content/10/3/R25)
  + 2012 Bowtie 2 - [Langmead and Salzberg](http://www.nature.com/nmeth/journal/v9/n4/full/nmeth.1923.htm)
  + 2009 BWA - [Li and Durbin](http://bioinformatics.oxfordjournals.org/content/25/14/1754.long)
  + 2010 BWA - [Li and Durbin](http://bioinformatics.oxfordjournals.org/content/26/5/589)
  + 2013 BWA-MEM - [Li](http://arxiv.org/abs/1303.3997)

Alignment relies on the reference genome being *indexed* so that the sequencing reads can be located more efficiently. The genome index is a highly-accessible data structure, and Galaxy includes indices for many popular genomes. 

#### 1.Align the example files  
